cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java

```diff
@@ -256,16 +256,25 @@ public void run() {
           handleAnomalyInProgress();
         } catch (InterruptedException e) {
           LOG.debug("Received interrupted exception.", e);
-          retryHandling = true;
-        } catch (KafkaCruiseControlException kcce) {
-          LOG.warn("Anomaly handler received exception when trying to fix the anomaly {}.", _anomalyInProgress, kcce);
-          retryHandling = true;
+          postProcessAnomalyInProgress = true;
+        } catch (OptimizationFailureException ofe) {
+          LOG.warn("Encountered optimization failure when trying to fix the anomaly {}.", _anomalyInProgress, ofe);
+          // If self-healing failed due to an optimization failure, that indicates a hard goal violation; hence there is
+          // no further processing anomaly detector can do without human intervention for the anomaly (i.e. other than
+          // what has already been done in the {@link #handlePostFixAnomaly(boolean, boolean, String)}).
+          postProcessAnomalyInProgress = false;
+        } catch (IllegalStateException ise) {
+          LOG.warn("Unexpected state prevents anomaly detector from handling the anomaly {}.", _anomalyInProgress, ise);
+          // An illegal state may indicate a transient process blocking self-healing (e.g. an ongoing execution not
+          // started by Cruise Control).
+          postProcessAnomalyInProgress = true;
```

Comment:
Looks like `IllegalStateException` is thrown when notifier reports un unsupported notification result [here](https://github.com/linkedin/cruise-control/blob/011b04aefec72cda142a0d067c9f943646915af1/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L311). In this case should  we set `postProcessAnomalyInProgress `  to false here.

I think for the scenario mentioned in comment(i.e. `an ongoing execution not started by Cruise Control`), we seems to already call `postProcessAnomalyInProgress` to handle [here](https://github.com/linkedin/cruise-control/blob/33a1fef9ee75ef4baff5552019a6fc8bda4b7db5/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L294).

Reply:
Good catch! We should ideally have another `Exception` type for ongoing execution exceptions, because `IllegalStateException` may be thrown from other code paths, as well -- but for now this seems like a reasonable assumption.
-- Created the following issue to address the issue better: https://github.com/linkedin/cruise-control/issues/1000